harvest-finance/src/mapping.ts

```diff
@@ -0,0 +1,213 @@
+import {AddVaultAndStrategyCall, DoHardWorkCall} from "../generated/HarvestEthController/HarvestEthController";
+import { Market } from "../generated/schema";
+import { ADDRESS_ZERO, getOrCreateAccount, getOrCreateERC20Token, getOrCreateMarket, investInMarket, redeemFromMarket, TokenBalance, updateMarket } from "./common";
+import { ProtocolName, ProtocolType } from "./constants";
+import { IERC20, Transfer } from "../generated/templates/RewardToken/IERC20";
+import { Address, BigInt, Bytes, ethereum, log } from "@graphprotocol/graph-ts";
+
+import { Vault } from '../generated/templates'
+import { Vault as VaultContract } from "../generated/templates/Vault/Vault";
+
+
+export function addVaultAndStrategy(add: AddVaultAndStrategyCall): void {
+  
+  let fAssetToken = IERC20.bind(add.inputs._vault);
+  let underlying = fAssetToken.try_underlying();
+  if(underlying.reverted) {
+    return;
+  }
+  let assetToken = IERC20.bind(underlying.value);
+  let farmToken = IERC20.bind(Address.fromString("0xa0246c9032bc3a600820415ae600c6388619a14d"));
+
+  // output
+  let fAssetTokenERC20 = getOrCreateERC20Token(add.block, add.inputs._vault);
+  fAssetTokenERC20.save();
+
+  // input
+  let assetERC20 = getOrCreateERC20Token(add.block, assetToken._address);
+  assetERC20.save();
+
+  // reward
+  let farmERC20 = getOrCreateERC20Token(add.block, farmToken._address);
+  farmERC20.save();
+  
+  let market = getOrCreateMarket(
+    add.block,
+    add.inputs._vault,
+    ProtocolName.HARVEST_FINANCE,
+    ProtocolType.LP_FARMING,
+    [assetERC20], // Asset
+    fAssetTokenERC20, // fAsset
+    [farmERC20], // FARM
+  )
+
+  market.save();
+
+  Vault.create(add.inputs._vault);
+}
+
+// update fAsset - mint/burn/transfer of users
+export function handleTransfer(event: Transfer): void {
```

Comment:
When mint or burn occurs market state changes as well - outputTokenBalance (fAsset amount) and inputTokenBalance(redeemable underlying asset) -> we need to have updateMarket executed